lib/ansible/modules/cloud/cloudstack/cs_storage_ip_range.py

```diff
@@ -0,0 +1,305 @@
+
+        ranges = self.cs.listStorageNetworkIpRange(**args)
+        if ranges:
+            for r in ranges['storagenetworkiprange']:
+                if ipv4_start == r['startip']:
+                    self.storage_ip_range = r
+                    break
+
+        return self.storage_ip_range
+
+    def present_storage_ip_range(self):
+        range = self.get_storage_ip_range()
+        if range:
+            range = self._update_storage_ip_range()
+        else:
+            range = self._create_storage_ip_range()
+        return range
+
+    def _create_storage_ip_range(self):
+        args = self._get_common_args()
+
+        self.result['changed'] = True
+
+        range = None
+        if not self.module.check_mode:
+            res = self.cs.createStorageNetworkIpRange(**args)
+            if 'errortext' in res:
+                self.module.fail_json(msg="Failed: '%s'" % res['errortext'])
+
+            poll_async = self.module.params.get('poll_async')
+            if poll_async:
+                range = self.poll_job(res, 'storagenetworkiprange')
+
+        self.result['changed'] = True
+        return range
+
+    def _update_storage_ip_range(self):
+        range = self.get_storage_ip_range()
+
+        args = self._get_common_args()
+        args['id'] = range['id']
+
+        if self.has_changed(args, range):
+            self.result['changed'] = True
+
+            if not self.module.check_mode:
+                res = self.cs.updateStorageNetworkIpRange(**args)
+                if 'errortext' in res:
+                    self.module.fail_json(msg="Failed: '%s'" % res['errortext'])
+
+                poll_async = self.module.params.get('poll_async')
+                if poll_async:
+                    range = self.poll_job(res, 'storagenetworkiprange')
+        return range
+
+    def absent_storage_ip_range(self):
+        range = self.get_storage_ip_range()
+        if range:
+            self.result['changed'] = True
+
+            args = {
+                'id': range['id'],
+            }
+            if not self.module.check_mode:
+                res = self.cs.deleteStorageNetworkIpRange(**args)
+                if 'errortext' in res:
+                    self.module.fail_json(msg="Failed: '%s'" % res['errortext'])
+
+                poll_async = self.module.params.get('poll_async')
+                if poll_async:
+                    success = self.poll_job(res, 'success')
+
+        return range
+
+    def get_pod(self, key=None):
+        pod = self.module.params.get('pod')
+        if not pod:
+            return None
+        args = {
+            'name': self.module.params.get('pod'),
+            'zoneid': self.get_zone(key='id'),
+        }
+        pods = self.cs.listPods(**args)
+        if pods:
+            return self._get_by_key(key, pods['pod'][0])
+        self.module.fail_json(msg="Pod %s not found in zone %s." % (self.module.params.get('pod'), self.get_zone(key='name')))
```

Comment:
just a little hint, there is a cloudstack optimized fail_json `self.fail_json` which would also return the zone on failure as return value. But this is fine as well